From 2e82c288d8bfe64fcaeca7662a3a854f88aa666f Mon Sep 17 00:00:00 2001
From: Xu Han <xu.han@intel.com>
Date: Wed, 24 Aug 2016 15:11:27 +0800
Subject: [PATCH 424/433] vgt: fix wrong stride value of sprite plane.

* Add sprint plane support besides primary plane.
* The stride value should read from stride register according to different plane type.
* This patch can fix MTBF instability bug #1096 on BDW platform.

Signed-off-by: Xu Han <xu.han@intel.com>
---
 drivers/gpu/drm/i915/vgt/fb_decoder.c |   19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/vgt/fb_decoder.c b/drivers/gpu/drm/i915/vgt/fb_decoder.c
index 7e886a1..1a473dd 100644
--- a/drivers/gpu/drm/i915/vgt/fb_decoder.c
+++ b/drivers/gpu/drm/i915/vgt/fb_decoder.c
@@ -219,11 +219,20 @@ int vgt_get_pixel_format_preskl(u32 plane_ctl,
 	return 0;
 }
 
-static u32 vgt_get_stride(struct vgt_device *vgt, int pipe, u32 tiled,
+static u32 vgt_get_stride(struct vgt_device *vgt, enum vgt_plane_type plane, int pipe, u32 tiled,
 			int stride_mask, int bpp)
 {
-	u32 stride_reg = __vreg(vgt, VGT_DSPSTRIDE(pipe)) & stride_mask;
-	u32 stride = stride_reg;
+	u32 stride_reg = 0;
+	u32 stride = 0;
+
+	if(plane == PRIMARY_PLANE)
+		stride_reg = __vreg(vgt, VGT_DSPSTRIDE(pipe)) & stride_mask;
+	else if(plane == SPRITE_PLANE)
+		stride_reg = __vreg(vgt, VGT_SPRSTRIDE(pipe)) & stride_mask;
+	else
+		vgt_warn("vgt_get_stride: unsupported plane:%d\n", plane);
+
+	stride = stride_reg;
 
 	if (IS_SKLPLUS(vgt->pdev)) {
 		switch (tiled) {
@@ -284,7 +293,7 @@ int vgt_decode_primary_plane_format(struct vgt_device *vgt,
 	memcpy(plane->drm_fmt_desc, com_plane_fmt.gen_pixel_format.desc, MAX_DRM_STR_SZ);
 
 	plane->base = __vreg(vgt, VGT_DSPSURF(pipe)) & GTT_PAGE_MASK;
-	plane->stride = vgt_get_stride(vgt, pipe, plane->tiled,
+	plane->stride = vgt_get_stride(vgt, PRIMARY_PLANE, pipe, plane->tiled,
 					com_plane_fmt.stride_mask, plane->bpp);
 
 	plane->width = (__vreg(vgt, VGT_PIPESRC(pipe)) & _PIPE_H_SRCSZ_MASK) >>
@@ -394,7 +403,7 @@ int vgt_decode_sprite_plane_format(struct vgt_device *vgt,
 	memcpy(plane->drm_fmt_desc, com_plane_fmt.gen_pixel_format.desc, MAX_DRM_STR_SZ);
 
 	plane->base = __vreg(vgt, VGT_SPRSURF(pipe)) & GTT_PAGE_MASK;
-	plane->stride = vgt_get_stride(vgt, pipe, plane->tiled,
+	plane->stride = vgt_get_stride(vgt, SPRITE_PLANE, pipe, plane->tiled,
 					com_plane_fmt.stride_mask, plane->bpp);
 
 	val = __vreg(vgt, VGT_SPRSIZE(pipe));
-- 
1.7.10.4

